As reported in #25533:

@WeiZhang555 unfortunately the commit does not fix the problem on s390x:
```
# git checkout origin/pr/25524
HEAD is now at 8541587... Bugfix: client hangs after run container
# TESTFLAGS='-check.f DockerSuite.TestEventsLimit'  \
       DOCKER_GRAPHDRIVER=vfs DOCKER_EXECDRIVER=native TIMEOUT="120m" \
       hack/make.sh gccgo test-integration-cli
+ go test -check.f DockerSuite.TestEventsLimit -check.v -check.timeout=120m -test.timeout=360m github.com/docker/docker/integration-cli
INFO: Testing against a local daemon
Cannot connect to the Docker daemon. Is the docker daemon running on this host?

Cannot connect to the Docker daemon. Is the docker daemon running on this host?
----------------------------------------------------------------------
FAIL: docker_cli_events_test.go:108: TestEventsLimit.pN52_github_com_docker_docker_integration_cli.DockerSuite

docker_cli_events_test.go:129:
    c.Assert(err, checker.IsNil, check.Commentf("%q failed with error", strings.Join(args, " ")))
... value *errors.errorString = &errors.errorString{s:"exit status 255: time=\"2016-08-09T16:24:04Z\" level=error msg=\"decode events error: unexpected EOF\" \n"} ("exit status 255: time=\"2016-08-09T16:24:04Z\" level=error msg=\"decode events error: unexpected EOF\" \n")
... "run --rm busybox true" failed with error

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (14.43s)
FAIL
---> Making bundle: .integration-daemon-stop (in bundles/1.13.0-dev/test-integration-cli)
+++++ cat bundles/1.13.0-dev/test-integration-cli/docker.pid
++++ kill 25560
/go/src/github.com/docker/docker/hack/make/.integration-daemon-stop: line 8: kill: (25560) - No such process
```

--- a/api/client/container/run.go
+++ b/api/client/container/run.go
@@ -8,7 +8,6 @@ import (
 	"runtime"
 	"strings"
 	"syscall"
-	"time"
 
 	"golang.org/x/net/context"
 
@@ -144,7 +143,6 @@ func runRun(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts *runOptions,
 		hostConfig.ConsoleSize[0], hostConfig.ConsoleSize[1] = dockerCli.GetTtySize()
 	}
 
-	startTime := time.Now()
 	ctx, cancelFun := context.WithCancel(context.Background())
 
 	createResponse, err := createContainer(ctx, dockerCli, config, hostConfig, networkingConfig, hostConfig.ContainerIDFile, opts.name)
@@ -221,6 +219,11 @@ func runRun(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts *runOptions,
 		})
 	}
 
+	statusChan, err := waitExitOrRemoved(dockerCli, context.Background(), createResponse.ID, hostConfig.AutoRemove)
+	if err != nil {
+		return fmt.Errorf("Error waiting container's exit code: %v", err)
+	}
+
 	//start the container
 	if err := client.ContainerStart(ctx, createResponse.ID, types.ContainerStartOptions{}); err != nil {
 		// If we have holdHijackedConnection, we should notify
@@ -233,9 +236,8 @@ func runRun(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts *runOptions,
 
 		reportError(stderr, cmdPath, err.Error(), false)
 		if hostConfig.AutoRemove {
-			if _, errWait := waitExitOrRemoved(dockerCli, context.Background(), createResponse.ID, hostConfig.AutoRemove, startTime); errWait != nil {
-				logrus.Debugf("Error waiting container's removal: %v", errWait)
-			}
+			// wait container to be removed
+			<-statusChan
 		}
 		return runStartContainerErr(err)
 	}
@@ -260,14 +262,7 @@ func runRun(dockerCli *client.DockerCli, flags *pflag.FlagSet, opts *runOptions,
 		return nil
 	}
 
-	var status int
-
-	// Attached mode
-	status, err = waitExitOrRemoved(dockerCli, ctx, createResponse.ID, hostConfig.AutoRemove, startTime)
-	if err != nil {
-		return fmt.Errorf("Error waiting container to exit: %v", err)
-	}
-
+	status := <-statusChan
 	if status != 0 {
 		return cli.StatusError{StatusCode: status}
 	}

--- a/api/client/container/utils.go
+++ b/api/client/container/utils.go
@@ -21,40 +21,44 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
 		panic("Internal Error: waitExitOrRemoved needs a containerID as parameter")
 	}
 
-	var exitCode int
 	statusChan := make(chan int)
-	exitChan := make(chan struct{})
-	detachChan := make(chan struct{})
-	destroyChan := make(chan struct{})
-	eventsErr := make(chan error)
+	exitCode := 125
 
-	// Start watch events
-	eh := system.InitEventHandler()
-	eh.Handle("die", func(e events.Message) {
-		if len(e.Actor.Attributes) > 0 {
-			for k, v := range e.Actor.Attributes {
-				if k == "exitCode" {
-					var err error
-					if exitCode, err = strconv.Atoi(v); err != nil {
-						logrus.Errorf("Can't convert %q to int: %v", v, err)
-					}
-					close(exitChan)
-					break
+	eventProcessor := func(e events.Message, err error) error {
+		if err != nil {
+			statusChan <- exitCode
+			return fmt.Errorf("failed to decode event: %v", err)
+		}
+
+		stopProcessing := false
+		switch e.Status {
+		case "die":
+			if v, ok := e.Actor.Attributes["exitCode"]; ok {
+				code, cerr := strconv.Atoi(v)
+				if cerr != nil {
+					logrus.Errorf("failed to convert exitcode '%q' to int: %v", v, cerr)
+				} else {
+					exitCode = code
 				}
 			}
+			if !waitRemove {
+				stopProcessing = true
+			}
+		case "detach":
+			exitCode = 0
+			stopProcessing = true
+		case "destroy":
+			stopProcessing = true
 		}
-	})
 
-	eh.Handle("detach", func(e events.Message) {
-		exitCode = 0
-		close(detachChan)
-	})
-	eh.Handle("destroy", func(e events.Message) {
-		close(destroyChan)
-	})
+		if stopProcessing {
+			statusChan <- exitCode
+			// stop the loop processing
+			return fmt.Errorf("done")
+		}
 
-	eventChan := make(chan events.Message)
-	go eh.Watch(eventChan)
+		return nil
+	}
 
 	// Get events via Events API
 	f := filters.NewArgs()
@@ -68,45 +72,8 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
 		return nil, fmt.Errorf("can't get events from daemon: %v", err)
 	}
 
-	go func() {
-		eventsErr <- system.DecodeEvents(resBody, func(event events.Message, err error) error {
-			if err != nil {
-				return fmt.Errorf("decode events error: %v", err)
-			}
-			eventChan <- event
-			return nil
-		})
-		close(eventChan)
-	}()
+	go system.DecodeEvents(resBody, eventProcessor)
 
-	go func() {
-		var waitErr error
-		if waitRemove {
-			select {
-			case <-destroyChan:
-				// keep exitcode and return
-			case <-detachChan:
-				exitCode = 0
-			case waitErr = <-eventsErr:
-				exitCode = 125
-			}
-		} else {
-			select {
-			case <-exitChan:
-				// keep exitcode and return
-			case <-detachChan:
-				exitCode = 0
-			case waitErr = <-eventsErr:
-				exitCode = 125
-			}
-		}
-		if waitErr != nil {
-			logrus.Errorf("%v", waitErr)
-		}
-		statusChan <- exitCode
-
-		resBody.Close()
-	}()
 	return statusChan, nil
 }
 

--- a/api/client/container/utils.go
+++ b/api/client/container/utils.go
@@ -3,7 +3,6 @@ package container
 import (
 	"fmt"
 	"strconv"
-	"time"
 
 	"golang.org/x/net/context"
 
@@ -16,16 +15,18 @@ import (
 	"github.com/docker/engine-api/types/filters"
 )
 
-func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, containerID string, waitRemove bool, since time.Time) (int, error) {
+func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, containerID string, waitRemove bool) (chan int, error) {
 	if len(containerID) == 0 {
 		// containerID can never be empty
 		panic("Internal Error: waitExitOrRemoved needs a containerID as parameter")
 	}
 
 	var exitCode int
+	statusChan := make(chan int)
 	exitChan := make(chan struct{})
 	detachChan := make(chan struct{})
 	destroyChan := make(chan struct{})
+	eventsErr := make(chan error)
 
 	// Start watch events
 	eh := system.InitEventHandler()
@@ -54,45 +55,59 @@ func waitExitOrRemoved(dockerCli *client.DockerCli, ctx context.Context, contain
 
 	eventChan := make(chan events.Message)
 	go eh.Watch(eventChan)
-	defer close(eventChan)
 
 	// Get events via Events API
 	f := filters.NewArgs()
 	f.Add("type", "container")
 	f.Add("container", containerID)
 	options := types.EventsOptions{
-		Since:   fmt.Sprintf("%d", since.Unix()),
 		Filters: f,
 	}
 	resBody, err := dockerCli.Client().Events(ctx, options)
 	if err != nil {
-		return -1, fmt.Errorf("can't get events from daemon: %v", err)
+		return nil, fmt.Errorf("can't get events from daemon: %v", err)
 	}
-	defer resBody.Close()
 
-	go system.DecodeEvents(resBody, func(event events.Message, err error) error {
-		if err != nil {
+	go func() {
+		eventsErr <- system.DecodeEvents(resBody, func(event events.Message, err error) error {
+			if err != nil {
+				return fmt.Errorf("decode events error: %v", err)
+			}
+			eventChan <- event
 			return nil
-		}
-		eventChan <- event
-		return nil
-	})
+		})
+		close(eventChan)
+	}()
 
-	if waitRemove {
-		select {
-		case <-destroyChan:
-			return exitCode, nil
-		case <-detachChan:
-			return 0, nil
+	go func() {
+		var waitErr error
+		if waitRemove {
+			select {
+			case <-destroyChan:
+				// keep exitcode and return
+			case <-detachChan:
+				exitCode = 0
+			case waitErr = <-eventsErr:
+				exitCode = 125
+			}
+		} else {
+			select {
+			case <-exitChan:
+				// keep exitcode and return
+			case <-detachChan:
+				exitCode = 0
+			case waitErr = <-eventsErr:
+				exitCode = 125
+			}
 		}
-	} else {
-		select {
-		case <-exitChan:
-			return exitCode, nil
-		case <-detachChan:
-			return 0, nil
+		if waitErr != nil {
+			logrus.Errorf("%v", waitErr)
 		}
-	}
+		statusChan <- exitCode
+
+		resBody.Close()
+	}()
+	return statusChan, nil
 }
 
 // getExitCode performs an inspect on the container. It returns

